scripts/calculate_udf_formulas.py

```diff
@@ -0,0 +1,473 @@
+) -> str | float | int:
+    """Fetch a value from a UDF placeholder."""
+    recursive = True if placeholder[0] == "_" else False
+
+    # Extract UDF names
+    udf_names = re.findall(r"'(.*?)'", placeholder)
+    assert udf_names, f"Could not extract UDF names from placeholder '{placeholder}'"
+
+    # Where to fetch UDF
+    if "inp" in placeholder:
+        assert art_in, "Input artifact not provided"
+        obj = art_in
+        obj_type = "input artifact"
+    elif "outp" in placeholder:
+        assert art_out, "Output artifact not provided"
+        obj = art_out
+        obj_type = "output artifact"
+    elif "step" in placeholder:
+        assert step, "Step not provided"
+        obj = step
+        obj_type = "step"
+
+    # Iterate across UDFs
+    for i, udf_name in enumerate(udf_names):
+        # How to fetch UDF
+        if recursive:
+            assert type(obj) is Artifact, (
+                "Recursive UDF references only allowed for artifacts"
+            )
+            val = fetch_last(obj, udf_name, include_current=True, on_fail=None)
+        else:
+            val = obj.udf.get(udf_name)
+
+        if val is None:
+            if i + 1 < len(udf_names):
+                continue
+            else:
+                udf_names_quoted = [f"'{i}'" for i in udf_names]
+                msg = (
+                    "Could not resolve any of "
+                    + f"UDFs {', '.join(udf_names_quoted)} "
+                    + f"for {obj_type} '{obj.type.name if 'step' in placeholder else obj.name}' "
+                    + f"({obj.id}). Skipping calculation."
+                )
+                logging.info(msg)
+                raise SkipCalculation()
+        else:
+            break
+
+    # Returned values will pass through eval, so strings need
+    # to be escaped to not be interpreted as variables
+    if isinstance(val, str):
+        val = f"'{val}'"
+
+    if len(udf_names) > 1:
+        # Only clarify which UDF was used if multiple ones were provided
+        logging.info(f"Resolved {placeholder} UDF '{udf_name}' to {val}")
+
+    assert isinstance(val, (int, str, float)), f"Unexpected type for val: {type(val)}"
+
+    return val
+
+
+class SkipCalculation(Exception):
+    """Custom exception for skipping calculations."""
+
+    def __init__(self):
+        super().__init__()
+
+
+def parse_formula(formula: str) -> tuple[str, list[str]]:
+    """From a formula string, extract UDF placeholders.
+
+    Return the formula string with placeholders replaced by curly brackets,
+    as well as a list of the placeholders.
+
+    Also perform various assertions to sanity check the input and prevent code injection.
+    """
+    logging.info(f"Parsing formula: {formula}")
+
+    # Explicate non-placeholders allowed in formula, to prevent code injection
+    allowed_functions = [
+        ng_ul,
+        nM,
+        fmol_to_ng,
+        ng_to_fmol,
+        ng_ul_to_nM,
+        nM_to_ng_ul,
+    ]
+    allowed_strings = [
```

Comment:
Why are these needed? Are they used outside of UDF names?

Reply:
In LIMS we see a lot of use of the fields `Concentration` and `Conc. Units`, but we need a fixed unit in order to define the calculations, so these utility functions will force a concentration taken from a unit-agnostic set of UDFs to have a specific unit, either keeping it as is or using a supplied size to convert it if needed.

Reply:
Sorry, that refers to the functions `ng_ul` and `nM`. The rest are the functions that are defined inside the `formulas` modules and are in turn used by `ng_ul` and `nM`. I thought it made sense to include them since they make up a basic operations that might be useful to call directly in formulas.

Reply:
So are these functions special somehow? They don't follow the same syntax as the others?

Reply:
They are simply python functions that will be kept and evaluated in `eval_rh`

Reply:
I still don't really get why they need their own "allowed_strings" variable for just those two. What am I missing?

Reply:
Ah, it's to allow for: `outp['Conc. Units'] = 'ng/ul'`

Reply:
Oh, my mistake - the strings are sometimes passed directly as arguments to ng_ul() or nM() rather than being read from e.g. the "Conc. Units" UDF